internal/attestation/crafter/materials/materials.go

```diff
@@ -43,9 +44,52 @@ type crafterCommon struct {
 	input  *schemaapi.CraftingSchema_Material
 }
 
-// uploadAndCraft uploads the artifact to CAS and crafts the material
+func uploadAndCraftFromBytes(ctx context.Context, input *schemaapi.CraftingSchema_Material, backend *casclient.CASBackend, filename string, artifactData []byte, h cr_v1.Hash, l *zerolog.Logger) (*api.Attestation_Material, error) {
```

Comment:
I think there too much duplication here.

What about creating function called `uploadAndCraft` that supports a `io.Reader`, then the from bytes will pass a buffer and the fromFile will pass an opened file?

Reply:
Much better! that was one of the reasons I marked this PR as draft. I want to keep refactoring.